server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldSyntheticWriterHelper.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.mapper.flattened;
+
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A helper class that reconstructs the field including keys and values
+ * mapped by {@link FlattenedFieldMapper}. It looks at key/value pairs,
+ * parses them and uses the {@link XContentBuilder} to reconstruct the
+ * original object.
+ *
+ * Example: given the following set of sorted key value pairs read from
+ * doc values:
+ *
+ * "id": "AAAb12"
+ * "package.id": "102938"
+ * "root.asset_code": "abc"
+ * "root.from": "New York"
+ * "root.object.value": "10"
+ * "root.object.items": "102202929"
+ * "root.object.items": "290092911"
+ * "root.to": "Atlanta"
+ *
+ * it turns them into the corresponding object using {@link XContentBuilder}.
+ * If, for instance JSON is used, it generates the following after deserialization:
+ *
+ * `{
+ *     "id": "AAAb12",
+ *     "package": {
+ *         "id": "102938"
+ *     },
+ *     "root": {
+ *         "asset_code": "abc",
+ *         "from": "New York",
+ *         "object": {
+ *             "items": ["102202929", "290092911"],
+ *         },
+ *         "to": "Atlanta"
+ *     }
+ * }`
+ *
+ */
+class FlattenedFieldSyntheticWriterHelper {
+    private final SortedSetDocValues dv;
+
+    FlattenedFieldSyntheticWriterHelper(final SortedSetDocValues dv) {
+        this.dv = dv;
+    }
+
+    void write(final XContentBuilder b) throws IOException {
+        BytesRef currKeyValue = dv.lookupOrd(dv.nextOrd());
+        String currKey = FlattenedFieldParser.extractKey(currKeyValue).utf8ToString();
+        String currValue = FlattenedFieldParser.extractValue(currKeyValue).utf8ToString();
+        List<String> currPrefix = prefix(currKey);
+        List<String> prevPrefix = Collections.emptyList();
+        final List<String> values = new ArrayList<>();
+        values.add(currValue);
+        for (int i = 1; i < dv.docValueCount(); i++) {
+            BytesRef nextKeyValue = dv.lookupOrd(dv.nextOrd());
+            String nextKey = FlattenedFieldParser.extractKey(nextKeyValue).utf8ToString();
+            List<String> nextPrefix = prefix(nextKey);
+            List<String> startPrefix = prefixDiff(currPrefix, commonPrefix(currPrefix, prevPrefix));
+            List<String> endPrefix = prefixDiff(currPrefix, commonPrefix(currPrefix, nextPrefix));
```

Comment:
Maybe we can add a method here that loads a key/value from doc values and at the same time creates a start/end prefix? I think this will remove some repetition here and then this method is more focussed on checking whether to write to XContentBuilder based on previous, current and next fields. 

Maybe something like:

```java
FlattenedValue readValue() throws IOException {
        BytesRef raw = dv.lookupOrd(dv.nextOrd());
        String key = FlattenedFieldParser.extractKey(raw).utf8ToString();
        String value = FlattenedFieldParser.extractValue(raw).utf8ToString();
        return new FlattenedValue(key, value, prefix(key), suffix(key));
    }

    record FlattenedValue(String key, String value, List<String> prefix, List<String> suffix) {}
```